### ISSUE TITLE
Add Assert.IsType<T>(object obj)

### ISSUE THREAD

**Issue by spottedmahn** [Microsoft/testfx#413]
## Description
xUnit has a nice Assert method: [`public static T IsType<T>(object @object)`](https://github.com/xunit/assert.xunit/blob/dbb0b75cc8fa9349d63a97d2af453d2cd568faaf/TypeAsserts.cs#L73) that is lacking from MS Test.

MS Test has an "old" method: `Assert.IsInstanceOfType(object, typeof(object))`.  This feels very .Net 1.1 w/o generics.

## xUnit Advantages

1. The xUnit method uses generics making it cleaner IMHO
1. The xUnit method returns the object (allows for clean one-liner)

## Code Comparison

```csharp
var badRequestResult = Assert.IsType<BadRequestObjectResult>(result);

//do stuff with badRequestResult
var blah = badRequestResult.Value;
...
```

vs

```csharp
var badRequestResult = result as BadRequestObjectResult;
Assert.IsInstanceOfType(badRequestResult, typeof(BadRequestObjectResult));

//do stuff with badRequestResult
var blah = badRequestResult.Value;
...
```

## Open to Adding?

What to do you think about extending the Assert API?

---

Reference: [SO Post: xUnit.net IsType Equivalent in MS Test That Returns Type](https://stackoverflow.com/questions/50025527/xunit-net-istype-equivalent-in-ms-test-that-returns-type/50026234)

**Comment by AbhitejJohn** [comment]
@spottedmahn : Would [this](https://github.com/Microsoft/testfx-docs/blob/master/RFCs/002-Framework-Extensibility-Custom-Assertions.md) help? It does have the specific example you referred to. I agree that this would be nice addition in-box though.

**Comment by spottedmahn** [comment]
Hi @AbhitejJohn - that is close!

It should return the strongly typed object if it is that type IMO.

```csharp
public static class SampleAssertExtensions
{
	public static T IsOfType<T>(this Assert assert, object obj)
	{
		if(obj is T result)
		{
			return result;
		}
		throw new AssertFailedException("Type does not match");
	}
}
```

**Comment by spottedmahn** [comment]
I like this btw `Assert.That.IsNotNull(animal).And.IsOfType<Cat>(animal)`! 👍⚡

**Comment by AbhitejJohn** [comment]
@spottedmahn : yeah, I like the chaining a lot too. Makes things easier to express and its readable. Agree on the strong name typing. You might want to add that into a custom extension in your code base or possibly also add that out as a library. I've created [one](https://www.nuget.org/packages/MSTest.Assert.Extensions/1.0.3-preview) for asserts over exception handling based off of this but I can imagine others extending this out for type checking/comparison too.

**Comment by spottedmahn** [comment]
Hey @AbhitejJohn - that's brilliant 💲... create an extension library that can move much faster than the full library 🌉.  Like VS Power Tools!

I'll put in PR for `public static T IsOfType<T>(this Assert assert, object obj)`!

**Comment by AbhitejJohn** [comment]
I actually just put that out, so I can go crazy with this extension point and also see if it works for consumers. From the download count, it does look useful. The core framework would still want to update itself with a few of these APIs that everyone would need just so there isn't an extra step to find and pull in another package but that's a choice for the framework now.

**Comment by spottedmahn** [comment]
This should do the trick: [Gist - Generates Markdown From VisualStudio XML documentation files](https://gist.github.com/formix/515d3d11ee7c1c252f92).

[Sample MD output](https://github.com/formix/MegaCityOne/blob/master/MegaCityOne/doc/api.md).

[Source](https://stackoverflow.com/a/28653158/185123).

Thoughts 💭?

**Comment by spottedmahn** [comment]
I think it is great, put something out there.  If it works and gains traction, put it in the core lib!

**Comment by pascalpfeil** [comment]
+1 on this, this would clean up so many of my tests.
